src/AutoMapper/Internal/FeatureCollectionBase.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace AutoMapper.Internal
+{
+    public class FeatureCollectionBase<TValue> : IEnumerable<KeyValuePair<Type, TValue>>
+    {
+        private readonly Lazy<ConcurrentDictionary<Type, TValue>> _features = new Lazy<ConcurrentDictionary<Type, TValue>>(() => new ConcurrentDictionary<Type, TValue>());
```

Comment:
Configuration doesn't need to be thread safe. Also Lazy just makes things harder to read. Maybe drop the readonly flag and replace the dictionary with a readonly one after Seal. If it isn't too complicated.

Reply:
The `Lazy` is added to not allocate extra memory for the `Features`-collection that will be created for each `IMappingExpression`/`TypeMap` if that not is used. Should I still remove the lazy?

the `_features` field is protected by the set on `this[]` to not allow changes if the collection is sealed so no reason to replace the inner dictionary.

Reply:
I understand the intent. It just doesn't seem useful. I think that replacing the dictionary with a readonly one and dropping the flag is a more elegant way to achieve the same purpose. But maybe it's just me :)

Reply:
I will change it but will not initiate the collection before first set/add-usage. then we don't allocate not needed memory.

Reply:
Allocating an empty collection doesn't really matter. Having extra code to avoid that is just premature optimization.